src/cpp/shared/xml/utils_xml.h

```diff
@@ -1,16 +1,92 @@
   tinyxml2::XMLError QueryCharText(tinyxml2::XMLElement *el, char *dest);
   void QueryText(tinyxml2::XMLElement *el, char *dest);
   void QueryText(tinyxml2::XMLElement *el, std::string &dest);
-
+  template<typename T>
+  void PushBack(tinyxml2::XMLDocument *doc, tinyxml2::XMLNode *dest, const char* name, const T &val) {
+    tinyxml2::XMLElement *elem = doc->NewElement(name);
+    elem->SetText(val);
+    dest->InsertEndChild(elem);
+  }
+  template <typename T>
+  void WriteArray(tinyxml2::XMLDocument *doc, tinyxml2::XMLNode *dest, const char* name, const T &src) {
+    for(int i = 0; i < ELEMENTS_IN(src); i++) {
+        tinyxml2::XMLElement *elem = doc->NewElement(name);
+        elem->SetAttribute("index", i);
+        elem->SetAttribute("value", src[i]);
+        dest->InsertEndChild(elem);
+      }
+  }
+  template <>
+  inline void WriteArray(tinyxml2::XMLDocument *doc, tinyxml2::XMLNode *dest, const char* name, const std::vector<int> &src) {
+    for(int i = 0; i < src.size(); i++) {
+        tinyxml2::XMLElement *elem = doc->NewElement(name);
+        elem->SetAttribute("index", i);
+        elem->SetAttribute("value", src[i]);
+        dest->InsertEndChild(elem);
+      }
+  }
+  template <>
+  inline void WriteArray(tinyxml2::XMLDocument *doc, tinyxml2::XMLNode *dest, const char* name, const std::map<int, std::string> &src) {
+    for(auto &i : src) {
+        tinyxml2::XMLElement *elem = doc->NewElement(name);
+        elem->SetAttribute("index", i.first);
+        elem->SetAttribute("value", i.second.c_str());
+        dest->InsertEndChild(elem);
+      }
+  }
+  template <>
+  inline void WriteArray(tinyxml2::XMLDocument *doc, tinyxml2::XMLNode *dest, const char* name, const std::map<int, int> &src) {
+    for(auto &i : src) {
+        tinyxml2::XMLElement *elem = doc->NewElement(name);
+        elem->SetAttribute("index", i.first);
+        elem->SetAttribute("value", i.second);
+        dest->InsertEndChild(elem);
+      }
+  }
+  template <>
+  inline void WriteArray(tinyxml2::XMLDocument *doc, tinyxml2::XMLNode *dest, const char* name, const std::map<int, SMACKER_VIDEOS> &src) {
+    for(auto &i : src) {
+        tinyxml2::XMLElement *elem = doc->NewElement(name);
+        elem->SetAttribute("index", i.first);
+        elem->SetAttribute("value", i.second);
+        dest->InsertEndChild(elem);
+      }
+  }
+  template <>
+  inline void WriteArray(tinyxml2::XMLDocument *doc, tinyxml2::XMLNode *dest, const char* name, const std::map<int, std::set<int>> &src) {
+    for(auto &i : src) {
+      for(auto val : i.second) {
+        tinyxml2::XMLElement *elem = doc->NewElement(name);
+        elem->SetAttribute("index", i.first);
+        elem->SetAttribute("value", val);
+        dest->InsertEndChild(elem);
+      }
+    }
+  }
+  template<typename T>
+  void WriteCampaignDDArray(tinyxml2::XMLDocument *doc, tinyxml2::XMLNode *dest, const char* name, const T &src) {
```

Comment:
This function is specifically tied to representation choices of the campaign, and should be grouped with that code.

Reply:
The point is - it's not related to .cmp file in any way
On the other side all of the code in campaign_xml.cpp is related to .cmp file

Reply:
I see your point.

What I'm saying is that all the other functions in this file are general-purpose, whereas this is special purpose, so this is not a good place for it.

Reply:
Moved it back to where it was